### ISSUE TITLE
keyboard overlaps home button/gesture bar/layout switcher

### ISSUE THREAD

**Issue by jm355** [dessalines/thumb-key#1187]
### Steps to reproduce

Open the keyboard

### Expected behavior

Keyboard is above gesture bar

### Actual behavior

Keyboard overlaps gesture bar
![Screenshot_20241210-183644](https://github.com/user-attachments/assets/442b4891-e571-49e6-a050-78f67103d318)

![Screenshot_20241210-183821](https://github.com/user-attachments/assets/1726ad36-0e23-4173-b179-ff64319cfeb6)

### version of the program

4.0.8

### Android version

15

### Device

pixel 8

### Acknowledgements

- [X] I have written a short but informative title.
- [X] I have updated the app to **[the latest version](https://github.com/dessalines/thumb-key/releases/latest)**.
- [X] I have searched the existing issues and this is a new one, **NOT** a  duplicate or related to another open issue.
- [X] This is not a question or a discussion, in which case I should have  gone to [lemmy.ml/c/thumbkey](https://lemmy.ml/c/thumbkey)
- [X] This is a **single** bug report, in case of multiple bugs I will open  a separate issue for each one  (they can always link to each other if related)
- [ ] I have admitted that I am a clown by having checked this box, as I have not read these acknowledgements. 🤡
- [X] I have filled out all of the requested information in this form.

**Comment by lucascoelho07** [comment]
Mine overlaps with the bottom buttons on type-split layout, really annoying...
There used to be a black bar housing the buttons on the botton in Android 13 if I'm remembering right.
![Image](https://github.com/user-attachments/assets/46a0ea4b-5ecd-4b39-b256-626cca1d7a26)

**Comment by dessalines** [comment]
What version / android device? I wonder why it would be ignoring the systemBarsPadding we added.

**Comment by lucascoelho07** [comment]
Thumb-Key version 4.0.35 on Xiaomi Redmi Note 11 with Android 15 (Matrixx Vanilla)

**Comment by jm355** [comment]
Happens on pixel fold as well

**Comment by dessalines** [comment]
@juschmitt would you be able to take a look at this? I'm super busy with other things atm.

**Comment by alerque** [comment]
I have three devices in with current Thumb Key versions on them right now: one has hardware back/etc. buttons, one uses a weird Android mod with no gesture bar (e-ink), but one is a phone device without buttons that uses the usual gesture bar as shown in screen shots here.

I just tested it it and I **cannot** reproduce this. The keyboard is nicely situated above the gesture bar. It is running Android 13 (LineageOS 20231002).

**Comment by lucascoelho07** [comment]
yeah, but mine also used to work back when I was using Android 13 LineageOS, it started when I switched to Android 15 Matrixx. Never tested Android 14 tho...

**Comment by alerque** [comment]
Well my Android 15 devices is actually older than my other once and has hardware butters rather than full frontal glass, so no gesture bar to test with...

**Comment by juschmitt** [comment]
I can reproduce this on my Pixel with Android 15 and 16.

For me these buttons only show up with the gesture navigation. 

For the globe button, it seems that setting

```
<input-method
    android:supportsSwitchingToNextInputMethod="false" android:isDefault="true">
    ...
</input-method>
```
in here:
https://github.com/dessalines/thumb-key/blob/9329eca4aa772afe5f2fd054ef3deb336e526c19/app/src/main/res/xml/method.xml#L3

should disable it. Which would be fine as the keyboard itself has this function too.

The other button to hide the keyboard seems to be a bit more complicated..

A quick search with ChatGPT suggests that we override this function 
```
override fun onComputeInsets(outInsets: InputMethodService.Insets?) {
    super.onComputeInsets(outInsets)
    outInsets?.contentTopInsets = 0
    outInsets?.visibleTopInsets = 0
}
```
which should move the button out of view. But maybe we don't want to get rid of that button completely and only move it slightly further down?

Anyways, I have not tested this, yet or implemented anything. Not sure when I could get to it though.

**Comment by alerque** [comment]
Does the position (center, corner), size, or backdrop setting of the keyboard have any effect on whether it overlaps?

**Comment by lucascoelho07** [comment]
No, the only one that can be somewhat of a workaround is the bottom offset, but it's still not perfect :(

**Comment by dessalines** [comment]
@juschmitt could you test the `android:supportsSwitchingToNextInputMethod="false" android:isDefault="true"`

and see if it fixes it?

**Comment by juschmitt** [comment]
I tested multiple things this morning and could not find a way to get rid of these two buttons.
The only solution that would remain in my eyes is adding an offset of `24.dp` to the bottom to at least not have them overlap with the keyboard.
This also seems to be what other IMEs like heliboard or florisboard do.

When testing it seemed to me that this is not an issue until Android 14. If no one else finds a better solution, I would suggest adding a bottom offset. I'll open a PR shortly for that.

**Comment by alerque** [comment]
> This also seems to be what other IMEs like heliboard or florisboard do.

In conjunction with my comment on #1463, is this really what all IMEs are having to do or is there something unique about these?

**Comment by juschmitt-ioki** [comment]
> > This also seems to be what other IMEs like heliboard or florisboard do.
> 
> In conjunction with my comment on [#1463](https://github.com/dessalines/thumb-key/pull/1463), is this really what all IMEs are having to do or is there something unique about these?

Right, see my comment [here](https://github.com/dessalines/thumb-key/pull/1463#issuecomment-3068565438). 

[This is also what other IMEs use](https://github.com/florisboard/florisboard/blob/0fa663f6582361985c098ac8087e9d10720f4119/app/src/main/kotlin/dev/patrickgold/florisboard/FlorisImeService.kt#L624). I just didn't catch that at first when looking into how others deal with that.

**Comment by juschmitt** [comment]
> > This also seems to be what other IMEs like heliboard or florisboard do.
> 
> In conjunction with my comment on [#1463](https://github.com/dessalines/thumb-key/pull/1463), is this really what all IMEs are having to do or is there something unique about these?

Right, see my comment [here](https://github.com/dessalines/thumb-key/pull/1463#issuecomment-3068595189). 

[This is also what other IMEs use](https://github.com/florisboard/florisboard/blob/0fa663f6582361985c098ac8087e9d10720f4119/app/src/main/kotlin/dev/patrickgold/florisboard/FlorisImeService.kt#L624). I just didn't catch that at first when looking into how others deal with that.

**Comment by alerque** [comment]
I just updated one of my Lineage OS devices from Android 13 to 15 and can confirm that *all* other things being equal what worked fine on Android 13 now shows me this same problem on Android 15 too.

**Comment by jm355** [comment]
Just to note, I can reproduce this on the pixel fold emulator in android studio, you just have to go into settings and enable gesture navigation. I don't have time to pull changes to test right now though

**Comment by alerque** [comment]
The "fix" for this turned out to not be right for all situations and was reverted in #1465, leaving this issue needing a resolution again.

**Comment by dessalines** [comment]
Thx for re-opening.

Anyone working on this, see my comment [here](https://github.com/dessalines/thumb-key/pull/1463#issuecomment-3073461227)

**Comment by jm355** [comment]
In terms of usability, I think having a weird gap on some devices is a less bad bug than the overlap issue. There is no practical problem there, such as accidentally closing the keyboard when you type a space, like there is for the overlap. In the meantime, can we maybe have a toggle in settings to select which padding number to use?

**Comment by alerque** [comment]
@dessalines I know our comments trying to narrow down what is actually going on are getting a bit spread out across issues (most recently in the closed PR #1463), but this issue is open so I'm continuing here.

I'm still trying to isolate steps to reproduce this issue (which I can do) *and* your version of the issue with bogus extra space (that bit I can't get any of my devices to trigger yet). Do you have any IMEs *besides* Thumb-Key enabled in your Android input settings? If none, can you enable one other IME (anything at all, even the system stock one). Does that trigger the extra bar as shown in this issue and fill in your blank space?

**Comment by dessalines** [comment]
@jm355 Is there any reason your device would be ignoring `navigationBarsPadding()`?

We can add an exception, but we need to know if its device specific, android version specific, etc.

**Comment by jm355** [comment]
It doesn't ignore it entirely, but it seems like that value offsets the nav bar but not the switch kb and minimize kb buttons 

on a pixel fold

small screen/closed
<img width="1080" height="2092" alt="Image" src="https://github.com/user-attachments/assets/0e622db7-fa19-451b-9496-5909045adf6d" />
<img width="1080" height="2092" alt="Image" src="https://github.com/user-attachments/assets/30d14530-fb3f-4ec3-8c72-d95b0c1c5637" />

Big screen/open
<img width="2208" height="1840" alt="Image" src="https://github.com/user-attachments/assets/f8618ba2-1896-46da-9fa1-3ed39cace2a3" />
<img width="2208" height="1840" alt="Image" src="https://github.com/user-attachments/assets/5b3c3a3f-c641-49c0-a4a9-30833d9263b8" />
<img width="1840" height="2208" alt="Image" src="https://github.com/user-attachments/assets/acb03511-42dd-4f07-82be-e7397055affc" />

**Comment by alerque** [comment]
As of the last 48 hours and for reasons unrelated to this project, I have acquired a fourth (used) hardware device. It came with a miserably bad stock ROM (Xioami Mi) at Android 11, and HUGE curved corners. For the record I hate the corners. In any case they are so huge that this problem is exacerbated from what I've seen before.

Quoting from #1463:

> I shouldn't have merged this, and am reverting it now. It adds a strange empty offset even on devices that don't use nav buttons (which is most devices now anyway)
> 
> ![Screenshot_2025-07-14-15-36-36-30_b3f38c9021993d9cc5da9a1388f25318](https://private-user-images.githubusercontent.com/930215/466170544-b6b254a2-5e6f-4ba7-afde-dccff588b177.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI4NDk1MjMsIm5iZiI6MTc1Mjg0OTIyMywicGF0aCI6Ii85MzAyMTUvNDY2MTcwNTQ0LWI2YjI1NGEyLTVlNmYtNGJhNy1hZmRlLWRjY2ZmNTg4YjE3Ny5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxOFQxNDMzNDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYWI4MzBiMmU5NTYxOTViZDIwYzY2NmY1ZjI3NTIxZmViOWU3YTk3YjljY2E0NzVhYTUzYzU5MjI3Mzg5ODgyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.TYIB8oJvMlW0cdNQjg2DGoLD8618wUhKs5Vc_IbGbD4)

I'm beginning to suspect that this screen shot does not actually show a different problem. The main issue here is (I suspect) that you have the keyboard backdrop turned off. The problem with this when paired with gesture nav is that it makes the Android default navigation helper bar transparent as well. When gesture nav is enabled this bar does not have the 3 nav buttons (back, home, menu) but it does still exist. On most screens it may just have a "hint" bar in it (the long horizontal bar suggesting users can pull up for home and app switcher). When an IME is active, it also has the "hide IME" button. In the event more than one language or keyboard is available, it will also have the language/keyboard switcher.

The only way I've found to disable the keyboard switcher is to only have one keyboard (and in the case of the AOSP or GBoard keyboards, only one language) activated. Then it goes away, but the "hide IME" button (down arrow) is still there.

I also suspect it is there on your screen, but since it is a small white label with no background (because you have the background disabled in thumb-key and it is mimicking that) you just can't see it.

Hence that space isn't bogus extra space, it is legitimate clearance for an unfortunately nearly invisible bar that is actually there.

Can you turn on the keyboard background and give it another look?

**Comment by juschmitt** [comment]
I have been looking into this again why there would be some bogus space.. but I think @alerque is on the right track with the last comment. 
I tried to reproduce the screenshot and got this on my Pixel 6:

![Screenshot_20250718-172055.png](https://github.com/user-attachments/assets/e300e446-a08e-4452-8ae7-0847abc35cde)

I circled the Hide IME button so you can see, that its nearly invisible with the settings. No background, light theme and color set to Pink.

**Comment by dessalines** [comment]
Using `navigationBarsPadding()`:

1 | 2
--- | ---
![Image](https://github.com/user-attachments/assets/e889d13e-30c1-4709-8231-0d20fbea8ff1) | ![Image](https://github.com/user-attachments/assets/e0e364d0-bd35-448c-ae95-bc0ba8479a28)

Correctly aligns to bottom, regardless of backdrop.

Using `safeDrawingPadding()`: 

1 | 2
--- | ---
![Image](https://github.com/user-attachments/assets/7fe640c2-b711-486d-8837-a770a5e026c2) | ![Image](https://github.com/user-attachments/assets/482ea32a-c96d-4b91-b7f7-64a3f900d8d9)

Adds a weird offset.

---

> Hence that space isn't bogus extra space, it is legitimate clearance for an unfortunately nearly invisible bar that is actually there.

[Based on this](https://developer.android.com/reference/kotlin/androidx/compose/foundation/layout/package-summary#(androidx.compose.foundation.layout.WindowInsets.Companion).safeDrawing()), I think you might be right. 

I'm willing to add a setting (default off), for `ignore_system_bottom_padding`, to account for my use case here, since I seem to be the odd one out.

**Comment by dessalines** [comment]
Just to clarify @juschmitt @jm355 , the `safeDrawingPadding()` (thumb-key v5.0.1) did work correctly?

**Comment by alerque** [comment]
Here are photos (because screen shots don't reveal what is going on with the curved corners) of 5.0.2 and 5.0.1 with equivalent settings on the same device:

<table>
<tr>
 <td><img alt="5.0.2" src="https://github.com/user-attachments/assets/5ea2f8a1-dc6a-453b-a0ed-db4f94a61920" />
 <td><img alt="5.0.1" src="https://github.com/user-attachments/assets/266ebe2d-0d91-4d03-86db-2599875bfb3c" />
</table>

**Comment by juschmitt** [comment]
Yes. For me it is working as I would have expected. Granted, I am using the Keyboard backdrop.
It does add another offset to the bottom, but thats better than having the Hide IME button overlap the keyboard. 

There would probably be a way to add a background that makes the gesture navigation artificially higher for the cases where the backdrop is disabled.
Not sure if that would make it better though 🤷

**Comment by jm355** [comment]
It worked for me, and I keep the backdrop disabled

**Comment by dessalines** [comment]
In that case I'll revert my revert, and then add a setting (default off) to not do `safeDrawingPadding()`
